src/cast_core/src/RunCommand.cpp

```diff
@@ -208,13 +208,135 @@ class DatabaseOperation {
     bool _awaitStepdown;
 };
 
+// Class to parse and hold the OnlyRunInInstance configuration. This option may be used to only
+// run under certain configurations Available configurations are specified in Options.
+class InstanceTypeFilter {
+public:
+    struct Options {
+        enum Code {
+            kStandalone,
+            kSharded,
+            kReplicaSet,
+            _size,          // Must before kNotInitialized
+            kNotInitialized  // Must be last
+        };
+        static constexpr std::string_view kNames[Code::_size] = {
+            [kStandalone] = "standalone",
+            [kSharded] = "sharded",
+            [kReplicaSet] = "replica_set"
+        };
+
+        static std::string getValidOptionsListAsString() {
+            std::stringstream listString;
+            for(int i = 0; i < Code::_size; ++i) {
+                if(i != 0) {
+                    listString << ", ";
+                }
+                listString << kNames[i];
+            }
+            return listString.str();
+        }
+
+        static std::optional<Code> stringToCode(std::string_view instanceType) {
+            for(int i = 0; i < Code::_size; ++i) {
+                if(kNames[i] == instanceType) {
+                    return static_cast<Code>(i);
+                }
+            }
+            return {};
+        }
+
+        static std::vector<Code> getCodesFromStrings(const std::vector<std::string>& options) {
+            std::vector<Code> codeOptions;
+            for(auto &option : options) {
+                if(auto code = stringToCode(option)) {
+                    codeOptions.push_back(*code);
+                } else {
+                    std::stringstream errorMsg;
+                    errorMsg << "OnlyRunInInstance or OnlyRunInInstances valid values are: "
+                        << Options::getValidOptionsListAsString();
+                    throw InvalidConfigurationException(errorMsg.str());
+                }
+            }
+            return codeOptions;
+        }
+    };
+
+    InstanceTypeFilter(PhaseContext &context, mongocxx::pool::entry& client) {
+        std::vector<std::string> stringOptions;
+        try {
+            stringOptions = context.getPlural<std::string>(
+                "OnlyRunInInstance", "OnlyRunInInstances", [&](const Node& node) {
+                    return node.to<std::string>();
+            });
+        } catch(const InvalidKeyException&) {
+            // Exception might be due to keys not found or other errors. If its due to keys not
+            // found ignore the expception. Otherwise rethrow.
+            if(context["OnlyRunInInstance"] || context["OnlyRunInInstance"]) {
```

Comment:
Should there be two different things on either side of this OR? Also, is this rethrow needed now that we're only catching the invalid key?

Reply:
You are right, one of those should be the plural, thanks! 

The `getPlural` uses the same InvalidKeyExpcetion for various cases. One of them is none of the keys (singular or plural) being specified, in which case we don't want to throw and proceed with default behaviour. The same exception is used to signal that both keys exist, which is incorrect and should throw. That is why the rethrow is needed.